torch_agents/cogment_verse_torch_agents/muzero/adapter.py

```diff
@@ -0,0 +1,801 @@
+        TrialConfig(
+            run_id=run_session.run_id,
+            environment=EnvironmentParams(
+                implementation=config.environment.implementation,
+                config=clone_config(
+                    config.environment.config,
+                    seed=config.environment.config.seed + i + config.training.demonstration_trials,
+                    render=False,
+                ),
+            ),
+            actors=[muzero_config, teacher_config],
+        )
+        for i in range(config.training.demonstration_trials)
+    ]
+
+    trial_configs = [
+        TrialConfig(
+            run_id=run_session.run_id,
+            environment=EnvironmentParams(
+                implementation=config.environment.implementation,
+                config=clone_config(
+                    config.environment.config,
+                    seed=config.environment.config.seed + i + config.training.demonstration_trials,
+                    render=False,
+                ),
+            ),
+            actors=[muzero_config],
+        )
+        for i in range(config.training.trial_count - config.training.demonstration_trials)
+    ]
+
+    return demonstration_configs + trial_configs
+
+
+async def single_agent_muzero_sample_producer_implementation(agent_adapter, run_sample_producer_session):
+    # allow up to two players for human/expert intervention
+    assert run_sample_producer_session.count_actors() in (1, 2)
+    state = None
+    step = 0
+    total_reward = 0
+    state, action, reward, policy, value = None, None, None, None, None
+    player_override = None
+
+    async for sample in run_sample_producer_session.get_all_samples():
+        observation = sample.get_actor_observation(0)
+        next_state = agent_adapter.tensor_from_cog_obs(observation)
+        done = sample.get_trial_state() == TrialState.ENDED
+        player_override = (
+            observation.current_player if observation.player_override == -1 else observation.player_override
+        )
+
+        if state is not None:
+            total_reward += reward
+
+            run_sample_producer_session.produce_training_sample(
+                (MuZeroSample(state, action, reward, next_state, done, policy, value), total_reward)
+            )
+
+        if done:
+            break
+
+        step += 1
+        state = next_state
+        action = agent_adapter.decode_cog_action(sample.get_actor_action(player_override))
+
+        if action < 0:
+            print("WARNING: override action is invalid, ignoring!")
+            action = agent_adapter.decode_cog_action(sample.get_actor_action(observation.current_player))
+
+        assert action >= 0
+
+        policy, value = agent_adapter.decode_cog_policy_value(sample.get_actor_action(observation.current_player))
+        reward = sample.get_actor_reward(player_override)
+
+
+def compute_targets(reward, value, reward_distribution, value_distribution):
+    with torch.no_grad():
```

Comment:
You can use the function decoration instead, right ?

Reply:
Fixed